tests/MongoDB.Driver.Tests/JsonDrivenTests/JsonDrivenRunOnThreadTest.cs

```diff
@@ -0,0 +1,141 @@
+
+        public JsonDrivenRunOnThreadTest(
+            JsonDrivenTestsContext testsContext,
+            IJsonDrivenTestRunner testRunner,
+            Dictionary<string, object> objectMap,
+            JsonDrivenTestFactory jsonDrivenTestFactory) : base(testsContext, testRunner, objectMap)
+        {
+            _jsonDrivenTestFactory = jsonDrivenTestFactory;
+        }
+
+        protected override void CallMethod(CancellationToken cancellationToken)
+        {
+            AssignTask(() => SubTest(false));
+        }
+
+        protected override Task CallMethodAsync(CancellationToken cancellationToken)
+        {
+            AssignTask(() => SubTest(true));
+            return Task.FromResult(true);
+        }
+
+        protected override void SetArgument(string name, BsonValue value)
+        {
+            switch (name)
+            {
+                case "operation":
+                    _operation = value.ToBsonDocument();
+                    return;
+            }
+
+            base.SetArgument(name, value);
+        }
+
+        // private methods
+        private void AssignTask(Action action)
+        {
+            if (_testContext.Tasks.ContainsKey(_name))
+            {
+                var taskAction = _testContext.Tasks[_name];
+                if (taskAction != null)
+                {
+                    throw new Exception($"Task {_name} must not be processed.");
+                }
+                else
+                {
+                    _testContext.Tasks[_name] = CreateTask(action);
+                }
+            }
+            else
+            {
+                throw new ArgumentException($"Task {_name} must be started before usage.");
+            }
+        }
+
+        private Task CreateTask(Action action)
+        {
+            return Task.Factory.StartNew(
+                action,
+                CancellationToken.None,
+                TaskCreationOptions.None,
+                new ThreadPerTaskScheduler());
+        }
+
+        private void SubTest(bool async)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(_operation, "name", "object", "arguments", "error");
+
+            var receiver = _operation["object"].ToString();
+            var name = _operation["name"].ToString();
+
+            var test = _jsonDrivenTestFactory.CreateTest(receiver, name);
+            test.Arrange(_operation);
+            if (async)
+            {
+                test.ActAsync(CancellationToken.None).GetAwaiter().GetResult();
+            }
+            else
+            {
+                test.Act(CancellationToken.None);
+            }
+            test.Assert();
+        }
+
+        // nested types
+        /// <summary>
+        /// https://code.msdn.microsoft.com/Samples-for-Parallel-b4b76364/sourcecode?fileId=44488&pathId=2098696067
+        /// Needs to call tasks as threads
+        /// </summary>
+        private class ThreadPerTaskScheduler : TaskScheduler
```

Comment:
Removed this class. Will check EG in the next patch

Reply:
Actually, it's still required. Otherwise there is no guaranties that operations will be called in parallel (I see on EG that concurrent tests fail from time to time (I debugged it before and it was related to the fact that thread steps were not called in parallel)).

Reply:
Previous Robert's comment: 
>Your links did not work for me. They open a page but they don't highlight any code so i don't know what you are trying to point at.

yes, Microsoft removed these samples from their site (between August and November 2019). However, I think it's still a good approach to use the task's advantages at the same time with guaranteed calling tasks in parallel

Reply:
we also discussed it previously https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/57#discussion_r332684768

Reply:
> Otherwise there is no guaranties that operations will be called in parallel 

Tasks run in parallel just like Threads do...

There is no guarantee that Threads will run in parallel either. Depends on how many threads are running and which gets scheduled when.

I remain unconvinced this scheduler is needed, and if it is that we have any sort of precise understanding of why.